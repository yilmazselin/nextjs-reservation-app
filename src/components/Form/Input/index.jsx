import { memo } from "react"
import styled from "styled-components"
import { font } from "../../../utils/helper"
import { colors, fontSize, fontWeight } from "../../../utils/styles"

const Container = styled.div`
  display: flex;
  flex-direction: column;
  position: relative;
  margin-bottom: ${(props) => props.marginBottom}px;
`
const Error = styled.p`
  position: absolute;
  bottom: -20px;
  ${font({
    size: fontSize.sm,
    weight: fontWeight.regular,
    color: colors.red
  })}
`

const Label = styled.label`
  ${font({
    size: fontSize.md,
    weight: fontWeight.regular,
    color: colors.gray
  })}
  order: 1;
  pointer-events: none;
  transform-origin: left top;
  transform: scale(1) translate3d(0, 22px, 0);
  transition: 200ms ease all;
  position: relative;
`

const StyledInput = styled.input`
  position: relative;
  border-radius: 0;
  display: flex;
  ${font({
    size: fontSize.md,
    weight: fontWeight.semiBold,
    color: colors.dark
  })}
  line-height: 24px;
  padding-bottom: 8px;
  outline: 0;
  border: 0;
  border-bottom: 1px solid;
  border-bottom-color: ${(props) => (props.error
    ? colors.red
    : colors.disabled)};
  flex: 1 1 auto;
  order: 2;
  transition: all 0.2s ease;
  &:focus ~ ${Label},&:not(:placeholder-shown) ~ ${Label} {
    color: #77869e;
    transform: scale(0.8) translate3d(0, -8px, 0);
  }
  &:not(:placeholder-shown) {
    border-bottom-color: ${colors.gray};
  }
  &:focus {
    border-bottom-color: ${(props) => (props.error
    ? colors.red
    : colors.green)};
  }
`

function Input({
  value,
  id,
  type = "text",
  label,
  name,
  onChange,
  marginBottom,
  error
}) {
  return <Container marginBottom={marginBottom}>
    <StyledInput
      error={error}
      value={value}
      id={id}
      type={type}
      name={name}
      onChange={onChange}
      placeholder=" "
    />
    <Label htmlFor={id}>{label}</Label>
    {error && <Error data-testid={`${name}-error`}>{error}</Error>}
  </Container>
}

export default memo(Input)
