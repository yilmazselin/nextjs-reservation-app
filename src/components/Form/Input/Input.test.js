import "@testing-library/jest-dom"
import { fireEvent, render, screen } from "@testing-library/react"
import "jest-styled-components"
import Input from "."

describe("<Input />", () => {
  it("Should render correctly", () => {
    render(<Input id="test-input" label="Test Input" />)
    const input = screen.getByLabelText("Test Input")

    expect(input).toMatchSnapshot()
  })

  it("Should call changeHandler onChange on input", () => {
    const mockHandler = jest.fn()

    render(<Input id="test-input" label="Test Input" onChange={mockHandler} />)
    const input = screen.getByLabelText("Test Input")

    fireEvent.change(input, {
      target: { value: "Message" }
    })

    expect(mockHandler.mock.calls.length).toBe(1)
  })
})
