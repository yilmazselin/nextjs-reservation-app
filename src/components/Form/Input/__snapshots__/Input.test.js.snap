// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> Should render correctly 1`] = `
.c0 {
  position: relative;
  border-radius: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: #121212;
  font-size: 16px;
  font-family: "NunitoSans-600","sans-serif";
  line-height: 24px;
  padding-bottom: 8px;
  outline: 0;
  border: 0;
  border-bottom: 1px solid;
  border-bottom-color: #BBC3CF;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  -webkit-order: 2;
  -ms-flex-order: 2;
  order: 2;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
}

.c0:not(:placeholder-shown) {
  border-bottom-color: #64738C;
}

.c0:focus {
  border-bottom-color: #0DAFC0;
}

<input
  class="c0"
  id="test-input"
  placeholder=" "
  type="text"
  value=""
/>
`;
